pkg/spanconfig/spanconfigreconciler/reconciler.go

```diff
@@ -0,0 +1,309 @@
+// construct). It does so incrementally and continuously, internally leveraging
+// SQL{Watcher,Translator}, KVAccessor, and Store to make progress.
+//
+// Let's walk through what it does. At a high-level, we maintain an in-memory
+// data structure that's up-to-date with the contents of the KV (at least the
+// subset of spans we have access to, i.e. the keyspace carved out for our
+// tenant ID). We watch for changes to SQL state (descriptors, zone configs),
+// translate the SQL updates to the flattened span+config form, "diff" the
+// updates against our data structure to see if there are any changes we need to
+// inform KV of. If so, we do, and ensure that our data structure is kept
+// up-to-date. We continue watching for future updates and repeat as necessary.
+//
+// There's only single instance of the Reconciler running for a given tenant at
+// a given point it time (mutual exclusion/leasing is provided by the jobs
+// subsystem). We needn't worry about contending writers, or the KV state being
+// changed from underneath us. What we do have to worry about, however, is
+// suspended tenants' not being reconciling while suspended. It's possible for
+// a suspended tenant's SQL state to be GC-ed away at older MVCC timestamps;
+// when watching for changes, we could fail to observe tables/indexes/partitions
+// getting deleted. Left as is, this would result in us never issuing a
+// corresponding deletion requests for the dropped span configs -- we'd be
+// leaving orphaned span configs lying around (taking up storage space and
+// creating pointless empty ranges). A "full reconciliation pass" is our attempt
+// to find all these extraneous entries in KV and to delete them.
+//
+// We can use our span config data structure here too, one that's pre-populated
+// with the contents of KV. We translate the entire SQL state into constituent
+// spans and configs, diff against our data structure to generate KV updates
+// that we then apply. We follow this with clearing out all these spans in our
+// data structure, leaving behind all extraneous entries to be found in KV --
+// entries we can then simply issue deletes for.
+//
+// TODO(irfansharif): The descriptions above presume holding the entire set of
+// span configs in memory, but we could break away from that by adding
+// pagination + retrieval limit to the GetSpanConfigEntriesFor API. We'd then
+// paginate through chunks of the keyspace at a time, do a "full reconciliation
+// pass" over just that chunk, and continue.
+//
+// TODO(irfansharif): We probably want some degree of batching when issuing RPCs
+// to KV -- right now we sent forth the entire set of updates since our last
+// checkpoint. For changes to, say, RANGE DEFAULT, the RPC request proto is
+// proportional to the number of schema objects.
+func (r *Reconciler) Reconcile(
+	ctx context.Context, checkpoint hlc.Timestamp, callback func(checkpoint hlc.Timestamp) error,
```

Comment:
My sense is this code could be a bit more decomposed if you didn't maintain all of the reconciliation state in closed over variables. The `Reconciler` to me is more of a factory that holds on to dependencies. It is itself stateless. Consider pulling out a struct that can hold the state so that various callbacks and what not can be methods.